katz/src/main/kotlin/katz/data/Validated.kt

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (C) 2017 The Katz Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package katz
+
+typealias ValidatedNel<E, A> = Validated<NonEmptyList<E>, A>
+
+sealed class Validated<out E, out A> : HK2<Validated.F, E, A> {
+
+    class F private constructor()
+
+    companion object {
+
+        fun <E, A> invalidNel(e: E): ValidatedNel<E, A> = Validated.Invalid(NonEmptyList(e, listOf()))
+
+        /**
+         * Converts a `Try<A>` to a `Validated<Throwable, A>`.
+         */
+        fun <A> fromTry(t: Try<A>): Validated<Throwable, A> = t.fold({ Invalid(it) }, { Valid(it) })
+
+        /**
+         * Converts an `Either<A, B>` to an `Validated<A, B>`.
+         */
+        fun <E, A> fromEither(e: Either<E, A>): Validated<E, A> = e.fold({ Invalid(it) }, { Valid(it) })
+
+        /**
+         * Converts an `Option<B>` to an `Validated<A, B>`, where the provided `ifNone` values is returned on
+         * the invalid of the `Validated` when the specified `Option` is `None`.
+         */
+        fun <E, A> fromOption(o: Option<A>, ifNone: () -> E): Validated<E, A> = o.fold(
+                { Invalid(ifNone()) },
+                { Valid(it) }
+        )
+    }
+
+    class Valid<out A>(val a: A) : Validated<Nothing, A>()
+
+    class Invalid<out E>(val e: E) : Validated<E, Nothing>()
+
+    fun <B> fold(fe: (E) -> B, fa: (A) -> B): B = when (this) {
+        is Valid -> fa(a)
+        is Invalid -> (fe(e))
+    }
+
+    val isValid = fold({ false }, { true })
+    val isInvalid = fold({ true }, { false })
+
+    /**
+     * Run the side-effecting function on the value if it is Valid
+     */
+    fun forEach(f: (A) -> Unit): Unit = fold({ }, f)
```

Comment:
:fire: let's stick to pure functional combinators for now until we have a need for these other non principled ones which can go in a separate lib as impure extensions